refinery/data_set_manager/views.py

```diff
@@ -1170,11 +1182,93 @@ def patch(self, request, uuid):
+                # start with source/start nodes and traverse down path
+                # over request file node and all other children
+                nodes_to_check = start_nodes
+                updated_nodes = []
+                while nodes_to_check:
+                    current_node = nodes_to_check.pop()
+                    if current_node not in updated_nodes:
+                        self.update_node_attribute(current_node,
+                                                   attribute_name,
+                                                   attribute_value)
+                        updated_nodes.append(current_node)
+                    children_nodes = current_node.children.all()
+                    for child_node in children_nodes:
+                        if child_node not in nodes_to_check:
+                            nodes_to_check.append(child_node)
+
+                return Response(
+                    NodeSerializer(node).data, status=status.HTTP_200_OK
+                )
+
+            return Response('Currently, you can only remove node files or '
+                            'edit non-derived files',
                             status=status.HTTP_405_METHOD_NOT_ALLOWED)
 
         return Response(uuid, status=status.HTTP_401_UNAUTHORIZED)
 
+    def get_start_nodes(self, children_nodes, start_nodes, checked_nodes):
+        """
+        traverses backwards from a list of nodes to grab all start nodes
+        :param children_nodes: list, beginning nodes to traverse from
+        :param start_nodes: list, start nodes for all the paths
+        :param checked_nodes: list, nodes who's parents have been traversed
+        :return: start_nodes
+        """
+        if children_nodes:
+            current_node = children_nodes.pop()
+            parent_nodes = current_node.parents.all()
+            for parent in parent_nodes:
+                if parent not in checked_nodes:
+                    children_nodes.append(parent)
+                    checked_nodes.append(parent)
+            if not parent_nodes and current_node not in start_nodes:
+                start_nodes.append(current_node)
+            self.get_start_nodes(children_nodes, start_nodes, checked_nodes)
+        return start_nodes
+
+    def update_node_attribute(self, node, solr_name, new_value):
+        """
+        updates node and it's annotated node's attribute value
+        :param node: obj, instance of a node object
+        :param solr_name: str, solr_name for the attribute to be updated
+        :param checked_nodes: str, value for attribute's value
+        :return: start_nodes
+        """
+        node_attributes = Attribute.objects.filter(node=node)
+        # splits solr name into type and subtype
+        attribute_obj = customize_attribute_response([solr_name])[0]
+        attribute_type = attribute_obj.get('attribute_type')
+        attribute_subtype = attribute_obj.get('display_name').lower()
+
+        if node_attributes.exists():
```

Comment:
For both of the filtering tasks here (`node_attributes ` & `annotated_nodes` below) I think you'd be fine attempting to iterate over the resulting querysets if they ended up being empty.  I think it may end up being clearer as to what we're trying to filter for if we include the desired fields in a single query.

```
node_attributes = Attribute.objects.filter(node=node, type=attribute_type, subtype=attribute_subtype)
for attribute in node_attributes:
    attribute.value = new_value
    attribute.save()
```

Reply:
can update for node attributes, but want to keep it for annotated_nodes as we don't want to update solr_index if there's no annotated_nodes